src/sentry/sentry_metrics/consumers/indexer/schema_validator.py

```diff
@@ -0,0 +1,51 @@
+import random
+from typing import Any, Optional
+
+from sentry_kafka_schemas.codecs import Codec, ValidationError
+
+from sentry import options
+from sentry.sentry_metrics.consumers.indexer.parsed_message import ParsedMessage
+
+
+class MetricsSchemaValidator:
+    """
+    MetricsSchemaValidator class implements schema validation. It takes into account whether
+    schema validation should be performed based on some heuristics rather than appling the
+    validation on all messages. This is done for performance reasons. It has been found during
+    performance testing that skipping schema validation can improve the performance of the
+    indexer by more than 100%. We want to be able to control this behavior with a configuration
+    option so that we can choose to do schema validation based of different considerations like
+      - How stable the use case is. If the use case is stable, we can choose to not do schema
+        validation
+      - How much we trust upstream to send valid messages. If we trust upstream to send valid
+        messages, we can choose to not do schema validation.
+      - How high the volume of data is. For high volume of data, we can choose to not do schema
+        validation.
+
+    Schema validation rules are followed in the following order:
+      - If the configuration option is not set for any use case, then schema validation is applied
+        to all messages of all use cases.
+      - If the configuratin option is not defined for a use case, then schema validation is applied
+        to all messages of that use case.
+      - If the configuration option is defined for a use case, then schema validation is applied to
+        messages of that use case based on the sample rate defined in the configuration option.
+    """
+
+    def __init__(self, input_codec: Optional[Codec[Any]], validation_option: Optional[str]) -> None:
+        self.input_codec = input_codec
+        self.validation_option = validation_option
+        if self.validation_option:
+            self.schema_validation_rules = options.get(self.validation_option, {})
```

Comment:
When and how many times does this class get initialized? If we are fetching the validation rule from options during `__init__`, we need to make sure that changing the option will immediately effect validation rate.

Reply:
This class would be initialized everytime the MessageProcessor's process_message function is called. That function gets called from [here](https://github.com/getsentry/sentry/blob/d4e132c80d1e4d9ff4efeb99eaaff68c9c0aaf09/src/sentry/sentry_metrics/consumers/indexer/parallel.py#L159). That means that the constructor would get called once for a batch of outer messages.

It is a deliberate choice to do this option lookup only once per outer batch. There is no reason to keep getting this value over and over for each individual message. The change in effect of the option should be really quick since the outer batches get created quite often.